library/proc_macro/src/diagnostic.rs

```diff
@@ -51,132 +100,125 @@ pub struct Diagnostic {
 
-    /// Returns the diagnostic `level` for `self`.
+    /// Creates a new note diagnostic with the provided message.
     #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-    pub fn level(&self) -> Level {
-        self.level
-    }
-
-    /// Sets the level in `self` to `level`.
-    #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-    pub fn set_level(&mut self, level: Level) {
-        self.level = level;
+    #[must_use = "diagnostics do nothing unless emitted"]
+    pub fn note(message: &str) -> Self {
+        Diagnostic {
+            level: Level::Note,
+            message: message.into(),
+            spans: vec![],
+            labels: vec![],
+            children: vec![],
+        }
     }
 
-    /// Returns the message in `self`.
+    /// Adds a help message to the diagnostic.
     #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-    pub fn message(&self) -> &str {
-        &self.message
+    pub fn with_help(mut self, message: &str) -> Self {
+        self.children.push(Self {
+            level: Level::Help,
+            message: message.into(),
+            spans: vec![],
+            labels: vec![],
+            children: vec![],
+        });
+        self
     }
 
-    /// Sets the message in `self` to `message`.
+    /// Adds a note message to the diagnostic.
     #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-    pub fn set_message<T: Into<String>>(&mut self, message: T) {
-        self.message = message.into();
+    pub fn with_note(mut self, message: &str) -> Self {
+        self.children.push(Self {
+            level: Level::Note,
+            message: message.into(),
+            spans: vec![],
+            labels: vec![],
+            children: vec![],
+        });
+        self
     }
 
-    /// Returns the `Span`s in `self`.
+    /// Adds one mark with the given `item.span()` to the last message. The mark
+    /// is "primary" if no other mark or label has been applied to the previous
+    /// message and "secondary" otherwise.
     #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-    pub fn spans(&self) -> &[Span] {
-        &self.spans
+    pub fn mark<S: Spanned>(mut self, item: S) -> Self {
+        self.last_message_mut().spans.push(item.span());
+        self
     }
 
-    /// Sets the `Span`s in `self` to `spans`.
+    /// Adds a spanned mark for every span in `item.spans()` to the last
+    /// message. The marks are "primary" if no other mark or label has been
+    /// applied to the previous message and "secondary" otherwise.
     #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-    pub fn set_spans<S: MultiSpan>(&mut self, spans: S) {
-        self.spans = spans.into_spans();
+    pub fn mark_all<S: Spanned>(mut self, item: S) -> Self {
+        let last_message = self.last_message_mut();
+        for span in item.spans() {
+            last_message.spans.push(span);
+        }
```

Comment:
Do we need `Extend`?

Reply:
I don't see why this wouldn't work, so I'll give it a shot.